_posts/2016-04-19-MIR.md

```diff
@@ -0,0 +1,725 @@
+expressions, and even method calls. Instead, those constructs are
+translated to a small set of primitives.
+
+The simple core language that MIR supports is not something you would
+want to program in. As we'll see, it makes things almost painfully
+explicit. But it's great if you want to write a type-checker or
+generate assembly code, as you now only have to handle the core
+operations that remain after MIR translation.
+
+To see what I mean, let's consider how we might simplify this fragment
+of code, which iterates over all the elements in a vector and
+processes them one by one:
+
+```rust
+for elem in vec {
+    process(elem);
+}
+```
+
+Rust itself offers three kinds of loops: `for` loops, like this one;
+`while` and `while let` loops, that iterate until some condition is
+met; and finally the simple `loop`, which just iterates until you
+break out of it. Each of these kinds of loops encapsulates a
+particular pattern, so they are quite useful when writing code. But
+for MIR, we'd like to reduce all of these into one core concept.
+
+A `for` loop in Rust works by converting a value into an iterator and
+then repeatedly calling `next` on that iterator. That means that we
+can rewrite the `for` loop we saw before into a `while let` loop
+that looks like this:
+
+```rust
+let mut iterator = vec.into_iter();
+while let Some(elem) = iterator.next() {
+    process(elem);
+}
+```
+
+By applying this rewriting, we can remove all `for` loops, but that
+still leaves multiple kinds of loops. So next we can imagine rewrite
+all `while let` loops into a simple `loop` combined with a `match`:
+
+```rust
+let mut iterator = vec.into_iter();
+loop {
+    match iterator.next() {
+        Some(elem) => process(elem),
+        None => break,
+    }
+}
+```
+
+We've already eliminated two constructs (`for` loops and `while`
+loops), but we can go further still. Let's turn from loops for a bit
+to look at the method calls that we see. In Rust, method calls like
+`vec.into_iter()` and `iterator.next()` are also a kind of
+[syntactic sugar][ss]. These particular methods are defined in traits,
+which are basically pre-defined interfaces.  For example, `into_iter`
+is a method in the `IntoIterator` trait.  Types which can be converted
+into iterators implement that trait and define how the `into_iter`
+method works for them. Similarly, `next` is defined in the `Iterator`
+trait. When you write a method call like `iterator.next()`, the Rust
+compiler automatically figures out which trait the method belongs to
+based on the type of the `iterator` and the set of traits in
+scope. But if we prefer to be more explicit, we could instead invoke
+the methods in the trait directly, using function call syntax:
+
+```rust
+// Rather than `vec.into_iter()`, we are calling
+// the function `IntoIterator::into_iter`. This is
+// exactly equivalent, just more explicit.
+let mut iterator = IntoIterator::into_iter(vec);
```

Comment:
It keeps the line shorter and doesn't have semantic implications, I think the point was method call desugaring, not path resolution.